src/protocol/slot.rs

```diff
@@ -0,0 +1,186 @@
+use std::io::{Read, Seek, SeekFrom, Write};
+
+use crate::nbt::Compound;
+use crate::protocol::{Decode, Encode, VarInt};
+
+/// Represents a slot in an inventory.
+#[derive(Clone, Default, Debug)]
+pub enum Slot {
+    #[default]
+    Empty,
+    Present {
+        item_id: VarInt,
+        item_count: u8,
+        nbt: Option<Compound>,
+    },
+}
+
+impl Slot {
+    /// Constructs a new slot.
+    pub const fn new(item_id: VarInt, item_count: u8, nbt: Option<Compound>) -> Self {
+        Self::Present {
+            item_id,
+            item_count,
+            nbt,
+        }
+    }
+
+    /// Constructs an empty slot.
+    pub const fn empty() -> Self {
+        Self::Empty
+    }
+
+    /// Returns `true` if there is an item present.
+    pub const fn is_present(&self) -> bool {
+        match self {
+            Slot::Empty => false,
+            Slot::Present { .. } => true,
+        }
+    }
+
+    /// Gets the item id.
+    ///
+    /// If the slot is empty, then `None` is returned.
+    pub fn item_id(&self) -> Option<&VarInt> {
+        match self {
+            Slot::Empty => None,
+            Slot::Present { item_id, .. } => Some(item_id),
+        }
+    }
+
+    /// Gets the item count.
+    ///
+    /// If the slot is empty, then `None` is returned.
+    pub fn item_count(&self) -> Option<&u8> {
+        match self {
+            Slot::Empty => None,
+            Slot::Present { item_count, .. } => Some(item_count),
+        }
+    }
+
+    /// Gets the item's nbt data.
+    ///
+    /// If the slot is empty or there is no nbt data, then `None` is returned.
+    pub fn nbt(&self) -> Option<&Compound> {
+        match self {
+            Slot::Empty => None,
+            Slot::Present { nbt, .. } => nbt.as_ref(),
+        }
+    }
+}
```

Comment:
If the `Slot` type is private to the protocol module it's fine to leave all this getter/setter stuff off and just make the fields public. (In fact, it already is because it's an `enum`). If we need these we can add them later.

Reply:
That makes sense, I'll remove these.